pkg/controller/nodes/array/handler.go

```diff
@@ -0,0 +1,593 @@
+							fmt.Sprintf("input arrays have different lengths: expecting '%d' found '%d'", size, collectionLength), nil),
+					), nil
+				}
+			}
+		}
+
+		if size == -1 {
+			return handler.DoTransition(handler.TransitionTypeEphemeral,
+				handler.PhaseInfoFailure(idlcore.ExecutionError_USER, errors.InvalidArrayLength, "no input array provided", nil),
+			), nil
+		}
+
+		// initialize ArrayNode state
+		maxAttempts := task.DefaultMaxAttempts
+		subNodeSpec := *arrayNode.GetSubNodeSpec()
+		if subNodeSpec.GetRetryStrategy() != nil && subNodeSpec.GetRetryStrategy().MinAttempts != nil {
+			maxAttempts = *subNodeSpec.GetRetryStrategy().MinAttempts
+		}
+
+		for _, item := range []struct {
+			arrayReference *bitarray.CompactArray
+			maxValue       int
+		}{
+			// we use NodePhaseRecovered for the `maxValue` of `SubNodePhases` because `Phase` is
+			// defined as an `iota` so it is impossible to programmatically get largest value
+			{arrayReference: &arrayNodeState.SubNodePhases, maxValue: int(v1alpha1.NodePhaseRecovered)},
+			{arrayReference: &arrayNodeState.SubNodeTaskPhases, maxValue: len(core.Phases) - 1},
+			{arrayReference: &arrayNodeState.SubNodeRetryAttempts, maxValue: maxAttempts},
+			{arrayReference: &arrayNodeState.SubNodeSystemFailures, maxValue: maxAttempts},
+		} {
+
+			*item.arrayReference, err = bitarray.NewCompactArray(uint(size), bitarray.Item(item.maxValue))
+			if err != nil {
+				return handler.UnknownTransition, err
+			}
+		}
+
+		// initialize externalResources
+		externalResources = make([]*event.ExternalResourceInfo, 0, size)
+		for i := 0; i < size; i++ {
+			externalResources = append(externalResources, &event.ExternalResourceInfo{
+				ExternalId:   buildSubNodeID(nCtx, i, 0),
+				Index:        uint32(i),
+				Logs:         nil,
+				RetryAttempt: 0,
+				Phase:        idlcore.TaskExecution_QUEUED,
+			})
+		}
+
+		// transition ArrayNode to `ArrayNodePhaseExecuting`
+		arrayNodeState.Phase = v1alpha1.ArrayNodePhaseExecuting
+	case v1alpha1.ArrayNodePhaseExecuting:
+		// process array node subNodes
+		currentParallelism := uint32(0)
+		messageCollector := errorcollector.NewErrorMessageCollector()
+		externalResources = make([]*event.ExternalResourceInfo, 0)
+		for i, nodePhaseUint64 := range arrayNodeState.SubNodePhases.GetItems() {
+			nodePhase := v1alpha1.NodePhase(nodePhaseUint64)
+
+			// do not process nodes in terminal state
+			if isTerminalNodePhase(nodePhase) {
+				continue
+			}
+
+			// create array contexts
+			arrayNodeExecutor, arrayExecutionContext, arrayDAGStructure, arrayNodeLookup, subNodeSpec, subNodeStatus, arrayEventRecorder, err :=
+				a.buildArrayNodeContext(ctx, nCtx, &arrayNodeState, arrayNode, i, &currentParallelism)
+			if err != nil {
+				return handler.UnknownTransition, err
+			}
+
+			// execute subNode
+			_, err = arrayNodeExecutor.RecursiveNodeHandler(ctx, arrayExecutionContext, arrayDAGStructure, arrayNodeLookup, subNodeSpec)
```

Comment:
the event stuff seems to be repeated, but i guess for array nodes it is slightly different right?

Reply:
exactly, there is a little bit of verbosity to keep this as a drop-in replacement. basically al this eventing stuff just captures everything that happens when evaluating subnodes and then aggregates the information into our `ExternalResources` field, which is how maptasks currently report everything.

I talked to the FE team briefly, for now I implemented this as a drop-in replacement, but maybe before GA we clean up this representation. very open to discussion here, but for now the goal was to make this as compatible as possible and provide a superset of functionality.